### DISCUSSION TITLE
Minimap

### DISCUSSION THREAD

**Discussion by JosephTLyons** [zed-industries/zed#23957] [Feature Requests]

**Comment by exalted** [comment]
FWIW, I am strongly against this 👎

**Comment by aexvir** [comment]
I'm also not a fan, it brings me zero value and it's distracting and takes space
but as long as it's optional I don't see why it couldn't be added
there is definitely people who likes it

**Comment by OldManFroggy** [comment]
best way I have seen a minimap implemented is -> inside a scroll bar, I think VSCode does it this way and as such takes up no extra UI space.

also a minimap should only show only selections, and lsp errors,warnings only in the minimap, changed lines of code (aka. git status) -> which makes minimap very functional and not just lipstick on a pig.

**Comment by aexvir** [comment]
@scott-ling the default vscode minimap? it works nothing like that for me 🤔 am I missing some config?
the only editor I recall showing the errors and warnings on the scrollbar is the intellij platform one but that's not really the minimap being discussed here; and usually git status indicators are on the left gutter, not on the minimap either

would you mind showing how the minimap looks for you on vscode? 😅 
I'm curious if I have been missing some nice fancy features all this time 😆

**Comment by OldManFroggy** [comment]
heres a little pic, just opened up a .bashrc and searched for an = sign, and you can see in the scrollbar there are all lines showing.

**Comment by aexvir** [comment]
can't see anything 😅 did the screenshot get attached?
but alright, I've reenabled it on vscode and I can indeed see the selections or search results in the minimap
I don't think this is very useful for me, especially seeing how much space the minimap takes
but yeah, I get your point

<img width="1506" alt="Screenshot 2022-07-14 at 21 52 13" src="https://user-images.githubusercontent.com/8055505/179071398-b1724604-81fa-4c52-bded-f6c080c9514a.png">

**Comment by OldManFroggy** [comment]
never said I was intelligent, seems github email drops images - ps. this is VSCode Insiders but think feature is on main branch also. 

See its in scroll bar -> not the mini map view setting which I never use, as stated above wastes soo much space :)

![image](https://user-images.githubusercontent.com/82799158/179071394-81ec65e5-9ea5-4eef-b827-8cad51b218b0.png)

**Comment by aexvir** [comment]
oh, hold up, that's something different... that's the scrollbar itself, which I have butchered with custom css 🤣 
yep, this is what I know from intellij based editors, but I wasn't aware vscode also had it
I turned off scrollbars looong time ago 😅 

yeah, this is definitely more useful 👍🏻

**Comment by OldManFroggy** [comment]
yeah, minimaps sux, minimap like features in scroll bar, awesome use of space

**Comment by JosephTLyons** [comment]
I'm dyslexic (being a dyslexic dev is a weird thing, but that's another story), so I often recognize code by the shape of it - which is where the minimap becomes useful for me.  I've never cared for the decorators in the scrollbar (JetBrains' IDEs) because I can't see the code's shape there, which is important to me.

**Comment by ernes-to** [comment]
I think the minimap by itself is not very useful. However, it's actually really nice when it shows red/yellow lines for ESLint and/or TypeScript errors/warnings throughout the entire file, allowing you to quickly click on that area of the minimap and scrolling you down to where the issue is.

**Comment by heyallan** [comment]
I also recognize code by it's shape. It's super useful. But I prefer seeing an outline of the document (like a table of contents). It gives me shape but also meaning in less space.

The breadcrumb feature already creates an outline, but only line by line, which is weird because if I am already at that line I already know where I am.

A document outline (as a form of minimap) would be useful for large documents.

Also, hints in the scrollbar are immediately useful.

**Comment by JosephTLyons** [comment]
@heyallan,

You're probably already aware of this, but in case you haven't seen this, we do have a `search buffer symbols` action that gives you a modal popup where you can see the overview of the current file's symbols.  You can then filter down to what you want and be taken to that point in the code.

<img width="854" alt="SCR-20221220-lup" src="https://user-images.githubusercontent.com/19867440/208762824-32f104d7-3976-4482-be59-8535bd6fd545.png">

**Comment by heyallan** [comment]
@JosephTLyons Exactly. The functionality is already there. I just would like to have it permanently open on the side as a form of minimap.

**Comment by natempv** [comment]
fwiw I love the minimap for giving the code a shape and body, helps me navigate it by feel, like a few people have mentioned already. like i know where things are from the topology of the thing, like it's a kaiju or a battleship. i've come to understand this is atypical but

also because I'll occasionally annotate my code with huge ascii/ansi art, kind of like xcode's //MARK comments but I usually just use figlet

**Comment by wzxu** [comment]
Also using the minimap to quickly navigate code by shape and colors. 
Would love a similar option to not render actual characters (makes the shape & color fuller and easier to scan).

<img width="462" alt="image" src="https://user-images.githubusercontent.com/5619348/229171672-b343527a-df8f-4c1f-abda-0e3c20f5b88a.png">

![image](https://user-images.githubusercontent.com/5619348/229174027-bf1d7775-d89d-4887-a2bd-3240e6dc0214.png)

**Comment by ojkelly** [comment]
I've found minimaps really useful for many of the reasons mentioned here. While I am missing them in Zed, solving for that might not require a minimap.

First, I rebound the outline to `cmd-o` to make it much quicker to open. The fact it's syntax highlighted, and searchable makes it _more useful than a minimap for moving around_ in a file.

```json
  {
    "bindings": {
      "cmd-o": "outline::Toggle"
    }
  }
```

However, I still find myself missing how vscode indicates problems in the file tree and minimap.

The problem multi-buffer is a really neat idea, but I don't find it useful for more complex problems. (The amount of padding and UI chrome can sometimes make things feel disconnected).

I think I could live without a minimap if the file tree and outline had a color/underline/dot/something to indicate a problem/warning.

I seem to be a very spatial programmer, in that I rarely search, and instead navigate to the file and part of the file when looking for something, or solving a bug. I think this is my brain indexing on where something is more than what it was named.

Given Zed's opportunity to rethink much of the IDE, and as much as I do love minimaps, I'm keen to see what you folks come up with.

**Comment by tsonubin** [comment]
> @heyallan,
> 
> You're probably already aware of this, but in case you haven't seen this, we do have a `search buffer symbols` action that gives you a modal popup where you can see the overview of the current file's symbols. You can then filter down to what you want and be taken to that point in the code.
> 
> <img alt="SCR-20221220-lup" width="854" src="https://user-images.githubusercontent.com/19867440/208762824-32f104d7-3976-4482-be59-8535bd6fd545.png">

This feature is really good on backend, but for the frontend code sometimes you have some annoying and long JSX function, in symbol view it looks like this. (Not very helpful) It will be so good if that symbol view can parse JSX!

<img width="842" alt="CleanShot 2023-04-07 at 13 45 19@2x" src="https://user-images.githubusercontent.com/10642396/230549001-d16fea41-c360-4086-84a6-4c0b3bb58c56.png">

**Comment by exalted** [comment]
> I seem to be a very spatial programming, in that I rarely search, and instead navigate to the file and part of the file when looking for something, or solving a bug. I think this is my brain indexing on where something is more than what it was named.

I love this insight!

**Comment by cktang88** [comment]
I think it would be nice to have configurable minimap, either in the scrollbar or standalone, since there's no clear consensus here.

**Comment by arguiot** [comment]
I really got used to minimaps. I'm not dyslexic but I do use it to recognise the shape of the code on long files. The other use case that I found SUPER useful is the `MARK` statements in Xcode.

**Comment by aeturnum** [comment]
Hey there - very excited about Zed - I wanted to echo others' comments about coming to rely on minimaps to navigate larger projects. At this point the lack of the option is a blocker on using adopting any new editor as a main editor.

I understand why people bring up `search buffer symbols` but I want to emphasize how the minimap feature allows you to spatially organize files by section and recognize layout. I both understand why some people don't care and want to encourage the Zed team to think about this as a core feature of a modern editor.

**Comment by fxsalazar** [comment]
Here's the alternative [Display hunks/diagnostics/search results in scroll bar](https://github.com/zed-industries/zed/issues/4866) that I'm personally looking forward to and kind of mentioned here https://github.com/zed-industries/zed/issues/5308
Let's 👍🏽  it up.

**Comment by JosephTLyons** [comment]
> Here's the alternative [Display hunks/diagnostics/search results in scroll bar](https://github.com/zed-industries/zed/issues/4866) that I'm personally looking forward to and kind of mentioned here https://github.com/zed-industries/zed/issues/5308
> 
> Let's 👍🏽  it up.

Git diff in scrollbar will be landing this next Wednesday, if all goes to plan.

**Comment by diocletiann** [comment]
Any chance of ticks in the scrollbar for bookmarks/errors/warnings/todo/etc like JetBrains? Even better if they were clickable.

What's the keybind for `search buffer symbols`? Can't find it. (edit: found it, "outline view").

**Comment by Congyuwang** [comment]
Xcode has one of the most useful minimap I've ever used:

You can use `// MARK: - XXX` to add section header, which will be displayed in the minimap, making it super easy to navigate around:
<img width="988" alt="截圖 2023-09-06 13 53 58" src="https://github.com/zed-industries/community/assets/52687642/18480c64-10b8-4e3e-b42b-7aec8e381409">

Putting the mouse over the minimap shows the name of the function the user is pointing at:
<img width="477" alt="截圖 2023-09-06 13 56 13" src="https://github.com/zed-industries/community/assets/52687642/bc1b5aa4-ac36-4126-9a03-21e713c9fe44">

**Comment by aranw** [comment]
I think one of the most useful features of the minimap that I'm missing in Zed is the ability to quickly see where symbols are being used 

![Screenshot 2023-09-26 at 09 27 38](https://github.com/zed-industries/community/assets/241565/c8136d23-150b-4fc5-a7bf-ff671cd31e84)

Edit: must admit with my current theme it is not the easiest to see. Be good if anyone has any recommendations for alternatives 😁

**Comment by fxsalazar** [comment]
Is not only about the theme, but having it in the scrollbar too...

**Comment by stonkol** [comment]
> best way I have seen a minimap implemented is -> inside a scroll bar, I think VSCode does it this way and as such takes up no extra UI space.
> 
> also a minimap should only show only selections, and lsp errors,warnings only in the minimap, changed lines of code (aka. git status) -> which makes minimap very functional and not just lipstick on a pig.

For sure, that is a clever solution I would like to see in Zed.

**Comment by normal-user-2** [comment]
Stopped using Zed only for these, these are killer features for me.

1. Showing red/yellow lines for errors/warnings throughout the `entire file` (through minimap)
1. Showing the filename's text color in `red` for errors/warnings throughout project `explorer`

These 2 options are very useful to find out the `errors` quickly, in stead of going to compiler.

I am hopeful to get back to Zed as soon as these 2 are in place. Enabling the notifications for this issue.

**Comment by antoniobrandao** [comment]
The minimap allows a faster scroll experience. Not only it is thicker than the scroolbar, but I can scroll half way through a long file in one mouse swipe.

As I code a file, I have a sense of what is in it. The minimap is a zoomed-out view that allows a particular type of navigation. I often click somewhere in the minimap to jump instantly to a section of the file, and I know that is the right place because I see the shape of the code in the minimap. 

It also helps me a lot with other things such as 
- Seeing other instances of an equal code fragment
- Seeing the duplicates mouses that land "offscreen" doing their things as I type
- Seeing the errors across the file in red at a glance (and their width (wether it's just a variable or a whole line))

I'm trying Zed today and I'm liking it. Hopefully this feature will arrive soon

**Comment by antoniobrandao** [comment]
> Also using the minimap to quickly navigate code by shape and colors. Would love a similar option to not render actual characters (makes the shape & color fuller and easier to scan).
> 
> <img alt="image" width="462" src="https://user-images.githubusercontent.com/5619348/229171672-b343527a-df8f-4c1f-abda-0e3c20f5b88a.png">
> 
> ![image](https://user-images.githubusercontent.com/5619348/229174027-bf1d7775-d89d-4887-a2bd-3240e6dc0214.png)

I think it only makes sense to render blocks. The text is never readable anyway. Blocks will always be more clear to the eye.

**Comment by nicholas-ochoa** [comment]
Just to add another example of usefulness, I haven't seen Sublime Text mentioned in this issue so far. When I find or highlight a symbol it highlights all instances of that symbol in the minimap which is very useful. Also, any lint / syntax errors and warnings are highlighted in the minimap:

https://github.com/zed-industries/zed/assets/126299/2a272681-64ec-4384-bfaa-cbef186ff2a6

**Comment by nazmul-hussain** [comment]
Minimap is so useful for so many reasons. Please add it as optional for those people who don't like it. But it is a really useful one as too many people asked above mentioned with usage. Someone said it is space-consuming! Do we write code across the entire area of our screen? If we split the screen and use multiple windows that could be a reason that it may take space otherwise it won't. In this case you can disable the feature

**Comment by zarifpour** [comment]
this might be crazy but... how about dragging the vertical scrollbar to the left to reveal, expand, and transform the scrollbar into a minimap. then when dragging it back to the right it collapses into the scrollbar as it is now.

after all, the minimap is just one big detailed scrollbar, isn't it?

**Comment by Moshyfawn** [comment]
I can't remember the specific software that had a similar type of interaction, but I think this is a great idea!

_P.S. Was it YouTube's timeline scrubbing that allowed for a more precise time selection when dragging up, that I'm thinking about?_

**Comment by joshuataylor** [comment]
Sorry to add to this, but a bit of context from other IDEs where this isn't the default:

One of the top 50 (~#32) plugins for IntelliJ/Jetbrains IDEs is [Code Glance Pro](plugins.jetbrains.com/plugin/18824-codeglance-pro), which adds this - just for some context for some massive "grain of salt" numbers.

**Comment by matthieuHenocque** [comment]
@Congyuwang

The `// MARK:` was just added to the March 2024 version of VSCode, as well as foldable `//#region` `//#endregion`

https://code.visualstudio.com/updates/v1_88

Zed's `Search buffer symbol` doesn't work in JavaScript Object, making it useless in JavaScript modules.

**Comment by tensei** [comment]
As a person with a disability who can only use a mouse (left and right click only), I rely heavily on the minimap in Visual Studio Code to scroll. It would be extremely helpful to have a similar feature in Zed as well.

**Comment by mocenigo** [comment]
> FWIW, I am strongly against this 👎

But it is optional. You do not have to use it if it is implemented!

**Comment by mocenigo** [comment]
> yeah, minimaps sux, minimap like features in scroll bar, awesome use of space
> […](#)
> On Jul 14 2022, at 8:56 pm, Alex Viscreanu ***@***.***> wrote: oh, hold up, that's something different... that's the scrollbar itself, which I have butchered with custom css 🤣 yep, this is what I know from intellij based editors, but I wasn't aware vscode also had it I turned off scrollbars looong time ago 😅 yeah, this is definitely more useful 👍🏻 — Reply to this email directly, view it on GitHub (https://github.com/zed-industries/feedback/issues/263#issuecomment-1184841051), or unsubscribe (https://github.com/notifications/unsubscribe-auth/ATXWUNQC3EHVLBCD3R4KLU3VUBWF5ANCNFSM53CW47CQ). You are receiving this because you were mentioned.

Again, it is personal. An editor should support it, because some people can jump to pieces of code just remembering the "shape", for instance. The users that do not want to use it need not activate it, and it would not be forced upon them.

**Comment by sleetdrop** [comment]
Besides xcode, the code reading tool SourceInsight 4 also implemented a minimap with helpful navigation functionalities. 
https://www.sourceinsight.com/doc/v4/userguide/Manual/Window_Tour/The_Overview_Scroller.htm

**Comment by antoniobrandao** [comment]
> // MARK:

@matthieuHenocque I love this! I used to create these large labels manually using ASCII generators. This is way more convenient.

**Comment by khanakia** [comment]
is there a way to enable minimap in zed

**Comment by danremollino** [comment]
Would love to have this. I frequently have longer stylesheets that I add ascii banners to for quick reference.

Would be even better if I could disable the default vertical scrollbar and use a minimap instead.

<img width="156" alt="minimap" src="https://github.com/zed-industries/zed/assets/11876088/484f0a49-c47a-4372-9eff-7383eaed989e">

**Comment by tjkohli** [comment]
+1 on this request. It really helps to scan & pinpoint getting to the correct code section especially in longer files.

**Comment by gabrielcosi** [comment]
+1

**Comment by ElhamAryanpur** [comment]
I would really love this as well, it helps get birdeye view of the code and locate blocks easily. +1

**Comment by insunaa** [comment]
A minimap is also useful when you have a multi-monitor setup and need to do a big jump-scroll. Currently with zed it's a fight to even hit the tiny scroll-bar handle when your document is a few thousand lines of code. 4px high and 10px wide on a 1440p monitor and if you miss it you either mark random stuff in your document, click into another application on your other monitor, jump to a random location in code or if you're truly unlucky hit the window resizer.

**Comment by pyromaniac** [comment]
I'm all for the minimap addition, this is an essential tool and helps a lot with code navigation and understanding. Please consider implementing it.

**Comment by jakobwildrain-si** [comment]
Is there a way to write it as an extension?
I love Zed and minimap is the only feature that I'm REALLY missing.

I would learn Rust to be able to write a minimap for Zed, but I don't think the way I would implement it would work for other people, so I'd rather write it as an extension, so if people don't like it they can pick another implementation of it.

**Comment by joseduartesi** [comment]
+1

**Comment by s-m-e** [comment]
For many reasons stated above (for me navigating large quantities of code, seeing where marked words and symbols show up, where text got replaced, where linters show issues etc) I am really missing this feature. It has been a great help to me in past. A GPU-accelerated map would open the door to a ton of cool use cases. If not implemented, I'd say it's a missed opportunity folks.

**Comment by Felipe-9** [comment]
I also use the `#MARK` and ascii banners features extensively on my documents, its very present on my workflow and i miss it very much.

**Comment by matthieuHenocque** [comment]
Came back here to share a bit of personal experience where the Minimap has been incredibly useful with the hope it will convince the Zed team to implement it.

We recently adopted Airbnb coding style at my company so I had to refactor huge JavaScript controllers from a legacy codebase. I used the Minimap's holistic point of view to evaluate the time needed to solve all of the linter's errors and warnings.

Regarding linters errors, it would be nice if there was a visual indicator of errors for things that are missing (like a semicolon) in the buffer. Just sayin'.

**Comment by mocenigo** [comment]
The argument is simple: Minimap is such a common feature nowadays — there are even some editors running in a terminal with a minimap — that it is simply needed to guarantee an editor’s wide adoption. And one of zed’s goals is to be widely adopted. 

This, and the ability to expose commands and command sequences in sidebar lists (like VScode), and two way sync with SyncTeX.

The latter could be achieved with proper scripting. Or a wider API, which then could be used by the extensions.

**Comment by gnat** [comment]
If you don't want it, don't enable it.

Nobody is forcing you to use it, but a lot of people do like this feature.

**Comment by nirui** [comment]
Minimap should be enabled by default.

Do you know how hard it is to use a scrollbar on a *large* file of 300 lines of code? Having to position my mouse cursor to be within that 20px wide of strip? It's like trying to climb a mountain with a noodle. It's 2024 and we must not continue this barbaricy.

**Comment by mocenigo** [comment]
> Minimap should be enabled by default.
> 
> Do you know how hard it is to use a scrollbar on a _large_ file of 300 lines of code? Having to position my mouse cursor to be within that 20px wide of strip? It's like trying to climb a mountain with a noodle. It's 2024 and we must not continue this barbaricy.

300 lines of code is a large file? What? :-) (Or maybe I do not get sarcasm.)

**Comment by SomeoneToIgnore** [comment]
There's 43 participants that get useless flood of e-mail and other notifications due to the new postings, hence I've locked the issue and urge people interested in the feature to submit PRs on the topic instead of _all this_ above.

Unfortunately, GH will prevent adding 👍 emojis for this too, but it's quite clear that the feature is wanted and in demand, it's just someone has to get the time to implement it.

I am not sure exactly how to implement this, but feel that all the bits are ready.
Here's the path I'd try:

* if ones opens any editor and zooms out to the end (e.g. by holding `cmd--` on mac), there will appear something that looks like a mini map already:

![image](https://github.com/user-attachments/assets/a2d23155-d633-4f4e-b6fc-31022cabc7ce)
* gpui allows absolute positioning of elements and various overflows, so we can take the buffer and wrap it into a new editor, enforce "max zoom out" property for it and absolute position it over everything in the top right corner modulo header and other menus above.
It feels that scrollbar and minimap should start at the same point above, and maybe even be of the same length?

* we might need to somehow ensure it's semi-transparent but it's not very needed for the first prototype if the whole minimap is disabled by default and can be enabled with a setting.
We should also make that new editor read-only, disable copilot suggestions, etc. and follow events of the "original" editor — so we can reflect selections and, maybe later, visible area border display.

* Also, we need to override click event handlers to scroll the original editor (and the minimap editor) if the minimap editor is clicked, same way the scrollbar does it.

*  Click through it and otherwise use it to ensure that it works normally and submit a PR.
